dataprocspawner/spawner.py

```diff
@@ -56,6 +58,95 @@ def url_path_join(*pieces):
   return result
 
 
+def _validate_proto(data, proto_cls):
+  """Utility method to strip unknown fields and enum values for proto parsing.
+
+  This method recursively calls _validate_proto_field for each field of the
+  proto, deleting unknown or invalid fields as it goes.
+
+  Args:
+  - dict data: representation of a proto as a (possibly nested) dict
+  - proto.message.MessageMeta proto_cls: class of the proto to which the data
+                                         should be parsed
+  Returns:
+    [String]: List of warnings of fields removed
+  """
+  if not isinstance(proto_cls, proto.message.MessageMeta):
+    return
+  if not data or not isinstance(data, dict):
+    return
+
+  meta = proto_cls._meta # pylint: disable=protected-access
+  warnings = []
+  fields_to_del = []
+  for field in data:
+    if field in meta.fields:
+      field_valid, new_warnings = _validate_proto_field(data[field], meta.fields[field])
+      warnings.extend(new_warnings)
+      if not field_valid:
+        warnings.append(f'Removing unknown/bad value {data[field]} for field {field}.')
+        fields_to_del.append(field)
+
+    else:
+      warnings.append(f'Removing unknown field {field} for class {proto_cls}')
+      fields_to_del.append(field)
+
+  for field in fields_to_del:
+    del data[field]
+
+  return warnings
+
+
+def _validate_proto_field(data, field_descriptor):
+  """Helper function for validating a single field of a proto.
+
+  *Only* validates message and enum fields.
+  Recursively calls _validate_proto for message fields.
+
+  Args:
+  Returns:
+    (Bool is_valid, [String] warnings):
+      is_valid: True if field is valid after this method returns. False if
+                field is still invalid after this method returns. If False,
+                the caller should either fail or remove offending field.
+      warnings: List of warnings of fields removed."""
+  if not data:
+    return True, []
+  fd = field_descriptor
+  warnings = []
+
+  if fd.message and isinstance(fd.message, proto.message.MessageMeta):
+    # Don't validate map fields
+    if fd.message._meta.options and fd.message._meta.options.map_entry: # pylint: disable=protected-access
+      return True, warnings
+
+    to_validate = [data] if not fd.repeated else data
+    for entry in to_validate:
+      warnings.extend(_validate_proto(entry, fd.message))
+    return True, warnings
+
+  elif fd.enum:
+    if fd.repeated:
+      # For repeated enum fields, filter out unknown values ourselves
+      to_del = []
+      for i, val in enumerate(data):
```

Comment:
Could we use enumerate(data.copy()) instead and use del data[val] to remove the need for to_del or would it break something?

Reply:
I think it would be difficult to do that here because we need to delete indices starting from the back. We could enumerate over the reversed list and then convert the indices back for the non-reversed original, but then that introduces more opportunities for indexing bugs.

I did apply this change to the method above, since deleting from dicts is easier.